flocker/node/agents/gce.py

```diff
@@ -35,6 +35,68 @@
 _METADATA_HEADERS = {u'Metadata-Flavor': u'Google'}
 
 
+def wait_for_operation(compute, operation, timeout):
+    """
+    Blocks until a GCE operation is complete, or timeout passes.
+
+    This function will then poll the operation until it reaches state
+    'DONE' or times out, and then returns the final operation resource
+    dict.
+
+    :param operation: A dict representing a pending GCE operation resource.
+        This can be either a zone or a global operation.
+    :param timeout: The amount of times in seconds to wait until timing out the
+        operation.
+
+    :returns dict: A dict representing the concluded GCE operation
+        resource or `None` if the operation times out.
+    """
+    operation_name = operation['name']
+    if 'zone' in operation:
+        # For zone operations, we need to extract the zone and the project.
+        # An example of a 'zone' value of a zone operation dict is:
+        #
+        # 'https://content.googleapis.com/compute/v1/'
+        # 'projects/<project-name>/zones/<zone-name>'
+        #
+        # Note that everything before 'projects' is sometimes not included.
+        zone_url_parts = operation['zone'].split('/')
+        project = zone_url_parts[-3]
+        zone = zone_url_parts[-1]
```

Comment:
Callers do know... but they just gave that information to a different call, and the information is available within the operation structure. So usably, it's not nice to have to keep specifying project and zone, and model wise it'd be a shame if we ever got a project/zone that was different than the project/zone that we could infer from the operation's structure.

Point taken about how unpleasant this is though, I think the unpleasantness comes from the brittleness of the parsing, and the terrible errors that it will produce if it guesses wrong.

I'll try to harden the parsing code and wrap key errors and index errors in a nicer error so that this can be considered to be more pleasant.